### PR TITLE
Allow remote login/logout from whitelisted origin

### PR DESCRIPTION
**Pull request by dtygel**
This PR (and riot-web https://github.com/vector-im/riot-web/pull/3949) adds an eventListener to MatrixChat component so that external referrer sites will be able to login or logout into matrix if the credentials and token are valid.
Only the same origin or allowed origin hostnames whitelisted in config.json through option "allowedPostMessageOrigins" will be able to successfully login/logout remotely.

Below is our specific use case:

USE CASE: We are developping a plugin to NextCloud (https://nextcloud.com/) in which we are integrating users using a common Ldap database.

After logging in into NextCloud, we are able to get the token and other credentials of the user from the matrix server. Now, when the user goes to a Nextcloud page that shows riot-web in an iframe, it's rather annoying to need to sign in again, since we already have the user's credentials and a valid token.

This PR solves the problem: the whitelisted caller sends a postMessage to the iframe, and the MatrixChat listens to message events. If the user is not logged in, the MatrixChat will check for the credentials and try to log in (using setLoggedIn method from LifeCycle).

### CODE REVIEW

**Comment by dbkr** [comment]
@matrixbot test this please

File: src/components/structures/MatrixChat.js [hunk 1]
```diff
@@ -299,6 +302,7 @@ module.exports = React.createClass({
         UDEHandler.stopListening();
         window.removeEventListener("focus", this.onFocus);
         window.removeEventListener('resize', this.handleResize);
+        window.removeEventListener("message", this.onMessage);
```
Comment (t3chguy) [inline src/components/structures/MatrixChat.js]: single quotes on new strings plox, as per code style

File: src/components/structures/MatrixChat.js [hunk 2]
```diff
@@ -255,6 +255,9 @@ module.exports = React.createClass({
         this.focusComposer = false;
         window.addEventListener("focus", this.onFocus);
 
+        // Listen to login credentials sent through postMessage from another site
+        window.addEventListener("message", this.onMessage);
```
Comment (t3chguy) [inline src/components/structures/MatrixChat.js]: see below

File: src/components/structures/MatrixChat.js [hunk 3]
```diff
@@ -893,6 +897,61 @@ module.exports = React.createClass({
...
+                : null;
+            var sameUser = (
+                currentCredentials &&
+                credentials.accessToken == currentCredentials.accessToken &&
+                credentials.homeserverUrl == currentCredentials.homeserverUrl &&
+                credentials.identityServerUrl == currentCredentials.identityServerUrl &&
+                credentials.userId == currentCredentials.userId
+            );
+            switch (credentials.action) {
+                case "login":
```
Comment (t3chguy) [inline src/components/structures/MatrixChat.js]: see above

File: src/components/structures/MatrixChat.js [hunk 4]
```diff
@@ -893,6 +897,61 @@ module.exports = React.createClass({
...
+                    console.log('postMessage: logging in from credentials sent by origin requestor');
+                    credentials.guest = false;
+                    if (sameUser) {
+                        console.log('postMessage: user is already logged in');
+                        return;
+                    }
+                    delete credentials.action;
+                    Lifecycle.setLoggedIn(credentials);
+                break;
+                case "logout":
```
Comment (t3chguy) [inline src/components/structures/MatrixChat.js]: see above

File: src/components/structures/MatrixChat.js [hunk 5]
```diff
@@ -893,6 +897,61 @@ module.exports = React.createClass({
...
+                    console.log('postMessage: logging out');
+                    if (sameUser) {
+                        Lifecycle.logout();
+                    } else {
+                        console.log("postMessage logout: user is not logged in or credentials are invalid");
```
Comment (t3chguy) [inline src/components/structures/MatrixChat.js]: see above

File: src/components/structures/MatrixChat.js [hunk 6]
```diff
@@ -893,6 +897,61 @@ module.exports = React.createClass({
...
+                case "login":
+                    console.log('postMessage: logging in from credentials sent by origin requestor');
+                    credentials.guest = false;
+                    if (sameUser) {
+                        console.log('postMessage: user is already logged in');
+                        return;
+                    }
+                    delete credentials.action;
+                    Lifecycle.setLoggedIn(credentials);
+                break;
```
Comment (t3chguy) [inline src/components/structures/MatrixChat.js]: indentation here is incorrect

File: src/components/structures/MatrixChat.js [hunk 7]
```diff
@@ -893,6 +897,61 @@ module.exports = React.createClass({
...
+                case "logout":
+                    console.log('postMessage: logging out');
+                    if (sameUser) {
+                        Lifecycle.logout();
+                    } else {
+                        console.log("postMessage logout: user is not logged in or credentials are invalid");
+                    }
+                break;
```
Comment (t3chguy) [inline src/components/structures/MatrixChat.js]: indentation is incorrect

**Comment by t3chguy** [comment]
doesn't the act of generating an AT create a deviceId, you don't seem to allow one to be passed in, is it implicitly generated by the server again in this case?
https://github.com/matrix-org/matrix-react-sdk/blob/1ff443956605ca818b212072767a1f6e515994b5/src/Login.js#L155

**Comment by dtygel** [comment]
@t3chguy : We can not make deviceId an obligatory field, since it's only needed in E2E cases, and Lifecycle.setLoggedIn doesn't demand it (#L307-309). So I added deviceId in the check if the sender sends it. ( last commit: https://github.com/matrix-org/matrix-react-sdk/pull/898/commits/454310e5181de90be38eef2553bd2f991cb57458 ).

Do you think this is fine?

**Comment by t3chguy** [comment]
only other thing is logout will delete someone's e2e keys without them accepting it, this this quite dangerous :/

**Comment by dtygel** [comment]
Yes @t3chguy . I agree with you. I was thinking of removing this possibility, but since Matthew proposed to have the whitelist barrier, we must agree that now somehow the sender must have some responsibility over using it. Maybe I should add this in README.md?

**Comment by t3chguy** [comment]
I think it should show the same dialog logout does, unless an extra param `forceDeleteKeys`? is also set?

**Comment by t3chguy** [comment]
that way you can have either functionality, force logout, or ask to logout

**Comment by dtygel** [comment]
Interstng.... Is there a method which calls the dialog before effectively doing the logout? (I'm checking now, but if you find it, I'll thank you ;-) )

**Comment by t3chguy** [comment]
Nope, you'd probably be best extracting the existing Dialog into an extension of BaseDialog in its own Class and then import it in both places

**Comment by t3chguy** [comment]
or just neaten it in where it is now and import it from there

**Comment by dtygel** [comment]
hmmm. Maybe create a specific file in components/views/dialogs/?

**Comment by t3chguy** [comment]
@dtygel thats what I mean when I said:
> Nope, you'd probably be best extracting the existing Dialog into an extension of BaseDialog in its own Class and then import it in both places

**Comment by ara4n** [comment]
@dtygel - thanks for putting this together :)  We're thinking about putting a generic postMessage interface into Riot in the near future (cc @rxl881), which this can form the basis of!

However, it's going to be critical to future-proof the API a bit.  Can I suggest that the `action` field is namespaced to `im.vector.login` and `im.vector.logout` or similar, to avoid it clashing with other possible methods we add in future?  Once standardised, these would become `m.login`, `m.logout` (yes, we're considering defining a postmessage API for matrix clients as an appendix of the matrix spec(!))

**Comment by dtygel** [comment]
@t3chguy : I added the boolean option `forceLogout` and now if it's not true (or not set) in postMessage, the user will be presented with the exportE2EKeysDialog before logging out. I then created specific docs about that as well.

**Comment by dtygel** [comment]
@t3chguy and @ara4n : Any news on this PR? Thanks! ;-)

**Comment by dtygel** [comment]
I've added a response from MatrixChat to the postMessage sender, so that it can be acknowledged that the login or logout request was succesfull. https://github.com/matrix-org/matrix-react-sdk/pull/898/commits/dff443dbd5c6153fcfea7fa1b052cb9dfe8a4b3c

**Comment by dtygel** [comment]
Following a suggestion from @lukebarnard1 , this PR now waits for the page view to be stable before doing anything with received postMessages from external origins. This is for preventing the concurrence of different login/logout processes at the same time.
So now the listener answers an object to the origin requestor with attributes _status_ and _msg_. _status_ can be `im.vector.error`, `im.vector.loading`, `im.vector.login` and `im.vector.logout`.
When the answer has _status_ `im.vector.error` or `im.vector.loading`, this means that the postMessage was not processed by riot-web.

**Comment by dtygel** [comment]
Should I keep updating this PR to solve conflicts with new versions in _develop_? Or is it out of Riot's roadmap?

**Comment by lukebarnard1** [comment]
Sorry that this has been dropped, @dtygel. I'll take a fresh look at this and see where you're at.

**Comment by lukebarnard1** [comment]
Actually it'd make far more sense if you were to resolve the conflicts first! Then I shall have a look.

**Comment by dtygel** [comment]
I hope all conflicts are done now, @lukebarnard1 ;-)

**Comment by dtygel** [comment]
It might be useful to use this "wrapper" to test the login from postMessage:

```
<?php
    $action = (isset($_GET['action'])) ? $_GET['action'] : "im.vector.login";
    $forceLogout = (isset($_GET['forceLogout'])) ? ', "forceLogout": true' : '';
?>
<!DOCTYPE HTML>
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
    <meta charset="utf-8" />
    <title>Dialog to Riot in an iframe</title>
    <script>
window.onload = function() {
    var destUrl = 'http://localhost:8080';
    var msg = {action: "<?= $action ?>"<?= $forceLogout ?>, accessToken:"{YOUR_VALID_ACCESS_TOKEN_HERE", homeserverUrl: "https://matrix.org", identityServerUrl: "https://vector.im", userId: "{YOUR_VALID_USER_ID_HERE}";
    var iframe = document.createElement('iframe');
    document.body.appendChild(iframe);
    iframe.src = destUrl;
    iframe.style = "position:absolute;margin:0;bottom:0;width:99%;height:80%";

    var refreshIntervalId;
    setTimeout(function(){
        refreshIntervalId = setInterval(function(){
            console.log('postMessage: sent login request');
	        iframe.contentWindow.postMessage(JSON.stringify(msg), "*");
        },500);
    },10);
    

    window.addEventListener('message',function(event) {
	    if(event.origin !== destUrl) return;
	    var data = JSON.parse(event.data);
	    console.log('postMessage: data received: '+JSON.stringify(data));
	    if(data.status != 'im.vector.loading') {
	        clearInterval(refreshIntervalId);
	    }
    },false);
};
    </script>
</head>
<body>
    <h1>Riot wrapper!</h1>
</body>
</html>
```

**Comment by lukebarnard1** [comment]
Removing my assignment as it's not on our immediate road map to add methods of login. 

(I'm also leaving the team this week, so I'll leave this in capable hands!)

**Comment by lukebarnard1** [comment]
:man_facepalming:

**Comment by jryans** [comment]
Thanks for making this contribution a while back. Since the code base has changed since this was opened, it no longer applies cleanly, and I don't think there's a need to keep it open in this state, as we can always find the code here again if needed. If you are still interested in pursuing this feature, please discuss with us in #riot-dev:matrix.org to find a good approach forward.

**Pull request closed without merging**